### PR TITLE
Tobikuhlmann/adjustment factor testing

### PR DESCRIPTION
**Pull request by ghost**
### Description

This PR adds test cases to the `EpochRewards.sol` function `updateTargetVotingYield()` to create confidence that the dynamic target voting yield adjustment works as intended. 

### Other changes

- Extended the testing documentation in packages/protocol/README.md
- Changed some epoch rewards init parameters to reflect more realistic values and testing

### Tested

All tests pass in the current configuration. You can test with `yarn run test epochrewards` (for only epoch rewards tests) and `yarn run test` (run all tests) in `packages/protocol/`.

Runtime for `yarn run test epochrewards` is around 4 minutes. 

### Backwards compatibility

Doesn't touch anything except testing, should be full backwards compatible. 

### Documentation

packages/protocol/README.md have been updated to extend the testing documentation

### CODE REVIEW

**Comment by ghost** [comment]
One weird thing I noticed was that the initialization parameters can not be chosen arbitrary as I expected. E.g. all test pass with a initialized total supply of 60 million and a reserve balance of 10 million. When increasing those numbers to e.g. 600 million total supply and 120 million reserve balance (`packages/protocol/tests/governance/network/epochrewards.ts` line 579 and 580), something in the test initialisation fails as the account where the funds are send from to the reserve does not have enough funds. Couldn't find where or how to change that, not sure if this is an issue and the initialisation should work independent on what numbers are chosen for total supply and reserve balance

File: packages/protocol/README.md [hunk 1]
```diff
@@ -109,3 +109,28 @@ yarn run test
...
+yarn run quicktest ${contract name}
+```
+
+For `quicktest` to work correctly a contract's migration dependencies have to be uncommented in `scripts/bash/backupmigrations.sh`.
+
+Compared to the normal test command, quicktest will:
+1. Not run the pretest script of building solidity (will still be run as part of truffle test) and compiling typescript. This works because truffle can run typescript "natively".
+2. Only migrate selected migrations as set in `backupmigrations.sh` (you'll likely need at least one compilation step since truffle seems to only run compiled migrations)
+
+
```
Comment (rcroessmann) [inline packages/protocol/README.md:136]: Love this addition! Very helpful.

File: packages/protocol/test/governance/network/epochrewards.ts [hunk 2]
```diff
@@ -64,9 +64,9 @@ contract('EpochRewards', (accounts: string[]) => {
   const nonOwner = accounts[1]
 
   const targetVotingYieldParams = {
-    initial: toFixed(new BigNumber(1 / 20)),
-    max: toFixed(new BigNumber(1 / 5)),
-    adjustmentFactor: toFixed(new BigNumber(1 / 365)),
+    initial: toFixed(new BigNumber(0.00016)),
+    max: toFixed(new BigNumber(0.0005)),
+    adjustmentFactor: toFixed(new BigNumber(0.00000112799)),
```
Comment (rcroessmann) [inline packages/protocol/test/governance/network/epochrewards.ts:69]: I am wondering how we can leave a hint on how to choose these parameters, especially the adjustmentfactor, in an economically reasonable way. How do you feel about scheduling some time in one of the next milestones to add a "parameter overview" section to docs.celo.org that explains the economic meaning of those parameters and links to background material like the doc you wrote up on the adjustmentFactor and its fallacies? We could then provide a link to those docs as a comment in the code.

File: packages/protocol/test/governance/network/epochrewards.ts [hunk 3]
```diff
@@ -657,6 +657,221 @@ contract('EpochRewards', (accounts: string[]) => {
...
+      })
+    })
+
+    // Maximum target voting yield is enforced
+    describe('When target voting yield is increased by adjustment factor', () => {
+      beforeEach(async () => {
+        const votingFraction = 0.1
+        const totalVotes = floatingSupply.times(votingFraction).integerValue(BigNumber.ROUND_FLOOR)
+        await mockElection.setTotalVotes(totalVotes)
+        // naive time travel: mining takes to long, just repeatedly update target voting yield. One call is one epoch travelled
```
Comment (rcroessmann) [inline packages/protocol/test/governance/network/epochrewards.ts]:
```suggestion
        // naive time travel: mining takes too long, just repeatedly update target voting yield. One call is one epoch travelled
```

Comment (rcroessmann) [reply packages/protocol/test/governance/network/epochrewards.ts]: Super nitty. Repeats multiple times further below.

Comment (rcroessmann) [reply packages/protocol/test/governance/network/epochrewards.ts]: What do you mean by "mining takes too long"?

File: packages/protocol/test/governance/network/epochrewards.ts [hunk 4]
```diff
@@ -657,6 +657,221 @@ contract('EpochRewards', (accounts: string[]) => {
...
+        const votingFraction = 0.3
+        const totalVotes = floatingSupply.times(votingFraction).integerValue(BigNumber.ROUND_FLOOR)
+        await mockElection.setTotalVotes(totalVotes)
+        // naive time travel: mining blocks takes to long, just repeatedly update target voting yield. One call is one epoch travelled
+        for (let i = 0; i < 5; i++) {
+          await epochRewards.updateTargetVotingYield()
+        }
+      })
+
+      it('target voting yield should be increased 5 times as expected', async () => {
```
Comment (rcroessmann) [inline packages/protocol/test/governance/network/epochrewards.ts:802]: How about adding another test that shows how a certain adjustmentfactor is supposed to effect the yield economically of 365 days? E.g. using the new value proposed in your CGP and checking for whether the 1 year change is as expected in the scenario used to derive this value.

File: packages/protocol/test/governance/network/epochrewards.ts (hunk 2 above)
Comment (ghost) [reply packages/protocol/test/governance/network/epochrewards.ts:69]: Good idea, that would be useful. Here is a ticket, feel free to edit: https://app.zenhub.com/workspaces/platform-economics---backlog-board-600a9f7399b43a001348ae1b/issues/celo-org/celo-monorepo/8578

File: packages/protocol/test/governance/network/epochrewards.ts (hunk 4 above)
Comment (ghost) [reply packages/protocol/test/governance/network/epochrewards.ts:802]: Do you imagine them in the 'per epoch' or 'annualized' world? I would recommend against adding tests for annualized yields here as everything in the code happens in the 'per epoch' world. I think the docs would be a good place to explain the annualization and epoch / yearly yield difference. 

I added two test cases below to test the 365 day change in the 'per epoch' world, that's a good idea

**Comment by ghost** [comment]
@rcroessmann didn't you add a comment about the naive time travel / mining? Can't see it anymore, maybe it got deleted with one of the two recent commits ... 

In the ganache testnet there is the option to time travel: Either increase the blocktime or 'mine' blocks into the future. I tried to use that to travel to future epochs first, however it took forever so I decided to do it the naive way and call the updateTargetVotingYield() function manually several times in a row

File: packages/protocol/test/governance/network/epochrewards.ts (hunk 2 above)
Comment (rcroessmann) [reply packages/protocol/test/governance/network/epochrewards.ts:69]: 🙏

File: packages/protocol/test/governance/network/epochrewards.ts (hunk 4 above)
Comment (rcroessmann) [reply packages/protocol/test/governance/network/epochrewards.ts:802]: Yeah definitely in epochs, I was a bit imprecise here. The test cases look great! But would probably rephrase some of the comments to "365 epochs (approximately 365 days, i.e. one year)".

Comment (ghost) [reply packages/protocol/test/governance/network/epochrewards.ts:802]: Good point, changed that

**Comment by rcroessmann** [comment]
> @rcroessmann didn't you add a comment about the naive time travel / mining? Can't see it anymore, maybe it got deleted with one of the two recent commits ...
> 
> In the ganache testnet there is the option to time travel: Either increase the blocktime or 'mine' blocks into the future. I tried to use that to travel to future epochs first, however it took forever so I decided to do it the naive way and call the updateTargetVotingYield() function manually several times in a row

I don't think I added something in this regard :) But repeatedly calling the `updateTargetVotingYield() ` should do in IMO.

**Review by rcroessmann — APPROVED** [review]
LGTM!

**Pull request merged**